src/core/coap/coap.hpp

```diff
@@ -55,12 +55,145 @@ namespace Coap {
  */
 
 /**
+ * Protocol Constants (RFC 7252).
+ *
+ */
+enum
+{
+    kAckTimeout                 = OPENTHREAD_CONFIG_COAP_ACK_TIMEOUT,
+    kAckRandomFactorNumerator   = OPENTHREAD_CONFIG_COAP_ACK_RANDOM_FACTOR_NUMERATOR,
+    kAckRandomFactorDenominator = OPENTHREAD_CONFIG_COAP_ACK_RANDOM_FACTOR_DENOMINATOR,
+    kMaxRetransmit              = OPENTHREAD_CONFIG_COAP_MAX_RETRANSMIT,
+    kNStart                     = 1,
+    kDefaultLeisure             = 5,
+    kProbingRate                = 1,
+
+    // Note that 2 << (kMaxRetransmit - 1) is equal to kMaxRetransmit power of 2
+    kMaxTransmitSpan            = kAckTimeout * ((2 << (kMaxRetransmit - 1)) - 1) *
+                                  kAckRandomFactorNumerator / kAckRandomFactorDenominator,
+    kMaxTransmitWait            = kAckTimeout * ((2 << kMaxRetransmit) - 1) *
+                                  kAckRandomFactorNumerator / kAckRandomFactorDenominator,
+    kMaxLatency                 = 100,
+    kProcessingDelay            = kAckTimeout,
+    kMaxRtt                     = 2 * kMaxLatency + kProcessingDelay,
+    kExchangeLifetime           = kMaxTransmitSpan + 2 * (kMaxLatency) + kProcessingDelay,
+    kNonLifetime                = kMaxTransmitSpan + kMaxLatency
+};
+
+/**
+ * This class implements metadata required for CoAP retransmission.
+ *
+ */
+OT_TOOL_PACKED_BEGIN
+class RequestMetadata
```

Comment:
After the merge, this name could be quite misleading as the metadata can be applied to both requests and responses. Perhaps it could be renamed to CoapMetadata or something like this?